tests/unit/utils_async_timer.cpp

```diff
@@ -9,141 +9,264 @@
+    EXPECT_TRUE(VerifyTimerExpiration(timer1));
+  }
+  {
+    AsyncTimer timer2{ToSeconds(kDefaultTimerDuration)};
+    EXPECT_TRUE(VerifyTimerExpiration(timer2));
+  }
+}
 
-TEST(AsyncTimer, DoubleWait) {
-  CheckTimeSimple();
-  CheckTimeSimple();
+TEST(AsyncTimer, RelativeTimingOrder) {
+  // Create timers with different durations to test ordering
+  AsyncTimer timer_short{ToSeconds(kShortTimer)};
+  AsyncTimer timer_medium{ToSeconds(kMediumTimer)};
+  AsyncTimer timer_long{ToSeconds(kLongTimer)};
+
+  // Wait for short timer (should expire first)
+  EXPECT_TRUE(WaitForExpiration(timer_short, kShortTimeout));
+  EXPECT_TRUE(timer_short.IsExpired());
+  EXPECT_FALSE(timer_medium.IsExpired());
+  EXPECT_FALSE(timer_long.IsExpired());
+
+  // Wait for medium timer (should expire second)
+  EXPECT_TRUE(WaitForExpiration(timer_medium, kMediumTimeout));
+  EXPECT_TRUE(timer_short.IsExpired());
+  EXPECT_TRUE(timer_medium.IsExpired());
+  EXPECT_FALSE(timer_long.IsExpired());
+
+  // Wait for long timer (should expire last)
+  EXPECT_TRUE(WaitForExpiration(timer_long, kLongTimeout));
+  EXPECT_TRUE(timer_short.IsExpired());
+  EXPECT_TRUE(timer_medium.IsExpired());
+  EXPECT_TRUE(timer_long.IsExpired());
 }
 
-TEST(AsyncTimer, MoveConstruct) {
-  const auto before = Now();
-  AsyncTimer timer_1{kIntervalInSeconds};
+TEST(AsyncTimer, MoveConstructor) {
+  AsyncTimer timer_1{ToSeconds(kDefaultTimerDuration)};
+
+  // Move construct timer_2 from timer_1
   AsyncTimer timer_2{std::move(timer_1)};
 
+  // After move, timer_1 is in moved-from state
+  // The implementation returns false for moved-from timers (null expiration_flag)
   EXPECT_FALSE(timer_1.IsExpired());
-  EXPECT_FALSE(timer_2.IsExpired());
-  const auto first_check_point = Now();
 
-  while (!timer_2.IsExpired()) {
-    ASSERT_LT(ElapsedMilis(before, Now()), 2 * kIntervalInMilis);
-  }
-  const auto second_check_point = Now();
+  // timer_2 should work normally
+  EXPECT_TRUE(VerifyTimerExpiration(timer_2));
 
+  // timer_1 should still return false (it's moved-from)
   EXPECT_FALSE(timer_1.IsExpired());
-  EXPECT_TRUE(timer_2.IsExpired());
-
-  EXPECT_LT(ElapsedMilis(before, first_check_point), kIntervalInMilis / 2);
-  EXPECT_NEAR(ElapsedMilis(before, second_check_point), kIntervalInMilis, kAbsoluteErrorInMilis);
 }
 
-TEST(AsyncTimer, MoveAssign) {
-  const auto before = Now();
-  AsyncTimer timer_1{2 * kIntervalInSeconds};
-  AsyncTimer timer_2{kIntervalInSeconds};
+TEST(AsyncTimer, MoveAssignment) {
+  AsyncTimer timer_1{ToSeconds(kMediumTimer)};  // Medium timer
+  AsyncTimer timer_2{ToSeconds(kShortTimer)};   // Short timer
 
+  // Initially both should not be expired
   EXPECT_FALSE(timer_1.IsExpired());
   EXPECT_FALSE(timer_2.IsExpired());
-  const auto first_check_point = Now();
 
+  // Move assign timer_1 to timer_2 (timer_2's original short timer is cancelled)
   timer_2 = std::move(timer_1);
+
+  // After move, timer_1 is moved-from, timer_2 has the medium timer
```

Comment:
why are we testing the state of moved-from object, we shouldn't make any assumptions about it

Reply:
FYI "we shouldn't make any assumptions about it" is a bad take. eg. moved from unique_ptr has a defined state, it will no longer hold onto the resource and it will be nullptr. Same here, moved from async timer is no longer responsible for observing the async event.